src_c/display.c

```diff
@@ -2028,17 +2082,281 @@ static PyObject *
 pg_toggle_fullscreen(PyObject *self, PyObject *args)
 {
     SDL_Window *win = pg_GetDefaultWindow();
-    int result;
+    int result, flags;
+    int window_w, window_h, w, h, window_display;
+    SDL_DisplayMode display_mode;
+    pgSurfaceObject *display_surface;
+    _DisplayState *state = DISPLAY_MOD_STATE(self);
+    GL_glViewport_Func p_glViewport = NULL;
+    SDL_SysWMinfo wm_info;
+    SDL_RendererInfo r_info;
+
     VIDEO_INIT_CHECK();
     if (!win)
         return RAISE(pgExc_SDLError, "No open window");
-    if (SDL_GetWindowFlags(win) & SDL_WINDOW_FULLSCREEN)
-        result = SDL_SetWindowFullscreen(win, SDL_WINDOW_FULLSCREEN);
-    else
-        result = SDL_SetWindowFullscreen(win, 0);
+
+    flags = SDL_GetWindowFlags(win) & SDL_WINDOW_FULLSCREEN_DESKTOP;
+    /* SDL_WINDOW_FULLSCREEN_DESKTOP includes SDL_WINDOW_FULLSCREEN */
+
+    SDL_VERSION(&wm_info.version);
+    SDL_GetWindowWMInfo(win, &wm_info);
+    if (pg_renderer !=NULL)
+        SDL_GetRendererInfo(pg_renderer, &r_info);
+
+    switch(wm_info.subsystem) {
+
+      // if we get this to work correctly with more systems, move them here
+      case SDL_SYSWM_WINDOWS:
+      case SDL_SYSWM_X11:
+#if SDL_VERSION_ATLEAST(2, 0, 2)
+      case SDL_SYSWM_WAYLAND:
+#endif
+	      break;
+
+      // These probably have fullscreen/windowed, but not tested yet.
+      // before merge, this section should be handled by moving items
+      // into the "supported" category, or returning early.
+      case SDL_SYSWM_COCOA: // we *need* to get this one to work
+
+#if SDL_VERSION_ATLEAST(2, 0, 3)
+      case SDL_SYSWM_WINRT: // currently not supported by pygame?
+#endif
+          break;
+
+      // On these platforms, everything is fullscreen at all times anyway
+      // So we silently fail
+      // In the future, add consoles like xbone/switch here
+      case SDL_SYSWM_DIRECTFB:
+      case SDL_SYSWM_UIKIT: // iOS currently not supported by pygame
+#if SDL_VERSION_ATLEAST(2, 0, 4)
+      case SDL_SYSWM_ANDROID: // currently not supported by pygame
+#endif
+          return PyInt_FromLong(-1);
+
+      // Untested and unsupported platforms
+#if SDL_VERSION_ATLEAST(2, 0, 2)
+    case SDL_SYSWM_MIR: //nobody uses mir any more, wayland has won
+#endif
+#if SDL_VERSION_ATLEAST(2, 0, 5)
+    case SDL_SYSWM_VIVANTE:
+#endif
+    case SDL_SYSWM_UNKNOWN:
+    default:
+	   return RAISE(pgExc_SDLError, "Unsupported platform");
```

Comment:
Raising an error here breaks backwards compatibility. Before you could call toggle_fullscreen and it would just do nothing. Instead, `return PyInt_FromLong(-1);` as above.

Reply:
I put this here mostly for testing purposes. This will be changed once the code has been tested on the supported platforms.